File: internal/config/tree.go
package config

import (
	"fmt"
	"os"
	"path/filepath"
	"sort"

	"gopkg.in/yaml.v3"
)

// ProjectNode represents one project in a potentially-nested ax hierarchy.
// Each node knows its own workspaces and its child projects so callers
// can render a tree without flattening workspace names.
type ProjectNode struct {
	Name                string // actual project name from the child config itself
	Alias               string // mount alias used by the parent children mapping
	Prefix              string // fully-qualified prefix used for merged names
	Dir                 string
	OrchestratorRuntime string
	Workspaces          []WorkspaceRef
	Children            []*ProjectNode
}

// WorkspaceRef is a workspace belonging to a project, with the merged
// "prefix.workspace" name that identifies its tmux session at runtime.
type WorkspaceRef struct {
	Name         string // original workspace name inside its project
	MergedName   string // fully-qualified name used by the daemon/tmux
	Runtime      string
	Description  string
	Instructions string
}

func (n *ProjectNode) DisplayName() string {
	if n == nil {
		return ""
	}
	if n.Alias == "" || n.Alias == n.Name {
		return n.Name
	}
	return fmt.Sprintf("%s (%s)", n.Alias, n.Name)
}

// LoadTree reads a config and recursively walks its children to produce a
// project tree. Unlike Load, it preserves the hierarchy instead of merging
// child workspaces into the parent's map.
func LoadTree(path string) (*ProjectNode, error) {
	if err := validateConfigTree(path); err != nil {
		return nil, err
	}
	seen := make(map[string]bool)
	return loadTreeRecursive(path, "", seen)
}

func loadTreeRecursive(path, prefix string, seen map[string]bool) (*ProjectNode, error) {
	absPath, err := filepath.Abs(path)
	if err != nil {
		return nil, fmt.Errorf("resolve config path: %w", err)
	}
	if seen[absPath] {
		return nil, fmt.Errorf("%w at %s", ErrCyclicChildren, absPath)
	}
	seen[absPath] = true
	defer delete(seen, absPath)

	data, err := os.ReadFile(absPath)
	if err != nil {
		return nil, fmt.Errorf("read config %s: %w", absPath, err)
	}
	var raw Config
	if err := yaml.Unmarshal(data, &raw); err != nil {
		return nil, fmt.Errorf("parse config %s: %w", absPath, err)
	}

	configDir := filepath.Dir(absPath)
	projectDir := configBaseDir(configDir)
	projectName := raw.Project
	if projectName == "" {
		projectName = filepath.Base(projectDir)
	}

	node := &ProjectNode{
		Name:                projectName,
		Prefix:              prefix,
		Dir:                 projectDir,
		OrchestratorRuntime: raw.OrchestratorRuntime,
	}

	// Workspaces defined directly in this project
	wsNames := make([]string, 0, len(raw.Workspaces))
	for name := range raw.Workspaces {
		wsNames = append(wsNames, name)
	}
	sort.Strings(wsNames)
	for _, name := range wsNames {
		ws := raw.Workspaces[name]
		merged := name
		if prefix != "" {
			merged = prefix + "." + name
		}
		node.Workspaces = append(node.Workspaces, WorkspaceRef{
			Name:         name,
			MergedName:   merged,
			Runtime:      ws.Runtime,
			Description:  ws.Description,
			Instructions: ws.Instructions,
		})
	}

	// Child projects
	childNames := make([]string, 0, len(raw.Children))
	for name := range raw.Children {
		childNames = append(childNames, name)
	}
	sort.Strings(childNames)
	for _, name := range childNames {
		child := raw.Children[name]
		childDir := resolveDir(projectDir, child.Dir)
		childPrefix := child.Prefix
		if childPrefix == "" {
			childPrefix = name
		}
		if prefix != "" {
			childPrefix = prefix + "." + childPrefix
		}

		childPath, err := ConfigPathInDir(childDir)
		if err != nil {
			continue
		}
		childNode, err := loadTreeRecursive(childPath, childPrefix, seen)
		if err != nil {
			if isStaleMissingChildError(err) {
				continue
			}
			return nil, wrapChildLoadError(name, childDir, err)
		}
		if childNode == nil {
			continue
		}
		childNode.Alias = name
		node.Children = append(node.Children, childNode)
	}

	return node, nil
}

File: rs/crates/ax-config/src/tree.rs
//! Recursive `Config` / `ProjectNode` loading.
//!
//! Mirrors `internal/config/config.go::Load` (workspace-merged view) and
//! `internal/config/tree.go::LoadTree` (hierarchical view). Stale child
//! references where the target `.ax/config.yaml` no longer exists are
//! skipped with a warning-like log, matching the Go behaviour so the rest
//! of the tree keeps loading.

use std::collections::{BTreeMap, BTreeSet};
use std::path::{Path, PathBuf};

use crate::paths::{config_path_in_dir, ConfigRoot};
use crate::schema::{Child, Config, LoadError, Workspace};

#[derive(Debug, thiserror::Error)]
pub enum TreeError {
    #[error(transparent)]
    Load(#[from] LoadError),
    #[error("cyclic ax children reference at {0}")]
    Cycle(PathBuf),
    #[error("load child {name} at {dir}: {source}")]
    Child {
        name: String,
        dir: String,
        #[source]
        source: Box<TreeError>,
    },
}

impl Config {
    /// Recursive load, merging child-project workspaces into the root
    /// `Config.workspaces` map with `prefix.name` keys.
    pub fn load(path: impl AsRef<Path>) -> Result<Self, TreeError> {
        let mut seen = BTreeSet::new();
        load_recursive(path.as_ref(), &mut seen)
    }

    /// Like [`Config::load`] but preserves the hierarchy instead of
    /// flattening workspaces.
    pub fn load_tree(path: impl AsRef<Path>) -> Result<ProjectNode, TreeError> {
        let mut seen = BTreeSet::new();
        load_tree_recursive(path.as_ref(), "", &mut seen)
    }
}

fn load_recursive(path: &Path, seen: &mut BTreeSet<PathBuf>) -> Result<Config, TreeError> {
    let abs = absolutize(path)?;
    if !seen.insert(abs.clone()) {
        return Err(TreeError::Cycle(abs));
    }

    let mut cfg = Config::read_local(&abs)?;
    initialize_local(&mut cfg, &abs);
    normalize_local(&mut cfg, &abs);

    let mut child_names: Vec<_> = cfg.children.keys().cloned().collect();
    child_names.sort();
    for name in child_names {
        let child = cfg.children.get(&name).cloned().unwrap_or_default();
        let Some(child_path) = config_path_in_dir(&child.dir) else {
            eprintln!(
                "warning: child {name:?} at {} has no config, skipping",
                child.dir
            );
            continue;
        };
        let child_cfg = match load_recursive(&child_path, seen) {
            Ok(v) => v,
            Err(e) if is_missing_file(&e) => {
                eprintln!(
                    "warning: child {name:?} at {} has no config, skipping",
                    child.dir
                );
                continue;
            }
            Err(e) => {
                return Err(TreeError::Child {
                    name,
                    dir: child.dir,
                    source: Box::new(e),
                });
            }
        };
        for (ws_name, ws) in child_cfg.workspaces {
            let merged = qualify_name(&child.prefix, &ws_name);
            if cfg.workspaces.contains_key(&merged) {
                eprintln!("warning: duplicate workspace {merged:?} from child {name:?}, skipping");
                continue;
            }
            cfg.workspaces.insert(merged, ws);
        }
    }

    seen.remove(&abs);
    Ok(cfg)
}

fn load_tree_recursive(
    path: &Path,
    prefix: &str,
    seen: &mut BTreeSet<PathBuf>,
) -> Result<ProjectNode, TreeError> {
    let abs = absolutize(path)?;
    if !seen.insert(abs.clone()) {
        return Err(TreeError::Cycle(abs));
    }

    let mut cfg = Config::read_local(&abs)?;
    initialize_local(&mut cfg, &abs);
    normalize_local(&mut cfg, &abs);

    let mut node = ProjectNode {
        name: cfg.project.clone(),
        alias: String::new(),
        prefix: prefix.to_owned(),
        dir: ConfigRoot::from_config_path(&abs).0,
        orchestrator_runtime: cfg.orchestrator_runtime.clone(),
        disable_root_orchestrator: prefix.is_empty() && cfg.disable_root_orchestrator,
        workspaces: Vec::new(),
        children: Vec::new(),
    };

    let mut ws_names: Vec<_> = cfg.workspaces.keys().cloned().collect();
    ws_names.sort();
    for name in ws_names {
        let ws = cfg.workspaces.get(&name).cloned().unwrap_or_default();
        let merged = qualify_name(prefix, &name);
        node.workspaces.push(WorkspaceRef {
            name,
            merged_name: merged,
            runtime: ws.runtime,
            description: ws.description,
            instructions: ws.instructions,
        });
    }

    let mut child_names: Vec<_> = cfg.children.keys().cloned().collect();
    child_names.sort();
    for name in child_names {
        let child = cfg.children.get(&name).cloned().unwrap_or_default();
        let child_prefix = qualify_name(prefix, &child.prefix);
        let Some(child_path) = config_path_in_dir(&child.dir) else {
            continue;
        };
        let child_node = match load_tree_recursive(&child_path, &child_prefix, seen) {
            Ok(v) => v,
            Err(e) if is_missing_file(&e) => continue,
            Err(e) => {
                return Err(TreeError::Child {
                    name,
                    dir: child.dir,
                    source: Box::new(e),
                });
            }
        };
        let mut child_node = child_node;
        child_node.alias = name;
        node.children.push(child_node);
    }

    seen.remove(&abs);
    Ok(node)
}

/// One project in the ax hierarchy — the tree form preserved by
/// `Config::load_tree`.
#[derive(Debug, Clone, Default)]
pub struct ProjectNode {
    pub name: String,
    /// Mount alias used by the parent's `children` map (empty at the root).
    pub alias: String,
    /// Fully-qualified prefix (e.g. `"team.sub"`) used when merging
    /// workspace names into the parent scope.
    pub prefix: String,
    pub dir: PathBuf,
    pub orchestrator_runtime: String,
    pub disable_root_orchestrator: bool,
    pub workspaces: Vec<WorkspaceRef>,
    pub children: Vec<ProjectNode>,
}

impl ProjectNode {
    #[must_use]
    pub fn display_name(&self) -> String {
        if self.alias.is_empty() || self.alias == self.name {
            self.name.clone()
        } else {
            format!("{} ({})", self.alias, self.name)
        }
    }
}

/// Workspace membership inside a [`ProjectNode`]. `merged_name` is the
/// name the daemon + tmux use at runtime.
#[derive(Debug, Clone, Default)]
pub struct WorkspaceRef {
    pub name: String,
    pub merged_name: String,
    pub runtime: String,
    pub description: String,
    pub instructions: String,
}

// ---------- helpers ----------

fn qualify_name(prefix: &str, name: &str) -> String {
    if prefix.is_empty() {
        name.to_owned()
    } else {
        format!("{prefix}.{name}")
    }
}

fn initialize_local(cfg: &mut Config, path: &Path) {
    if cfg.project.is_empty() {
        let root = ConfigRoot::from_config_path(path).0;
        let base = root.file_name().and_then(|os| os.to_str()).unwrap_or("");
        base.clone_into(&mut cfg.project);
    }
}

fn normalize_local(cfg: &mut Config, path: &Path) {
    let project_dir = ConfigRoot::from_config_path(path).0;

    let fallback_effort = cfg.codex_model_reasoning_effort.trim().to_owned();
    let workspace_names: Vec<_> = cfg.workspaces.keys().cloned().collect();
    for name in workspace_names {
        if let Some(ws) = cfg.workspaces.get_mut(&name) {
            ws.dir = resolve_dir(&project_dir, &ws.dir);
            if ws.codex_model_reasoning_effort.trim().is_empty() {
                ws.codex_model_reasoning_effort.clone_from(&fallback_effort);
            }
        }
    }

    let child_names: Vec<_> = cfg.children.keys().cloned().collect();
    for name in child_names {
        if let Some(child) = cfg.children.get_mut(&name) {
            child.dir = resolve_dir(&project_dir, &child.dir);
            if child.prefix.is_empty() {
                child.prefix.clone_from(&name);
            }
        }
    }
    let _ = Workspace::default(); // keep the re-export live to users
}

fn resolve_dir(base: &Path, value: &str) -> String {
    let value = if value.is_empty() { "." } else { value };
    if let Some(rest) = value.strip_prefix("~/") {
        if let Some(home) = std::env::var_os("HOME") {
            return Path::new(&home).join(rest).to_string_lossy().into_owned();
        }
    }
    let p = Path::new(value);
    if p.is_absolute() {
        value.to_owned()
    } else {
        base.join(p).to_string_lossy().into_owned()
    }
}

fn absolutize(path: &Path) -> Result<PathBuf, TreeError> {
    if path.is_absolute() {
        return Ok(path.to_path_buf());
    }
    let cwd = std::env::current_dir().map_err(|e| LoadError::Read {
        path: path.display().to_string(),
        source: e,
    })?;
    Ok(cwd.join(path))
}

fn is_missing_file(err: &TreeError) -> bool {
    if let TreeError::Load(LoadError::Read { source, .. }) = err {
        source.kind() == std::io::ErrorKind::NotFound
    } else {
        false
    }
}

#[allow(dead_code)]
fn _keep_children_type_live(_m: &BTreeMap<String, Child>) {}
